### PR TITLE
[C++] [Qt5]  Add initial version of File upload and download for Qt5 client

### LINKED ISSUE #3651: Qt5 client code fail to be compiled for https://github.com/jdegre/5GC_APIs/raw/master/TS29502_Nsmf_PDUSession.yaml
#### Bug Report Checklist

- [x] Have you provided a full/minimal spec to reproduce the issue?
- [ ] Have you validated the input using an OpenAPI validator ([example](https://apidevtools.org/swagger-parser/online/))?
- [ ] What's the version of OpenAPI Generator used?
- [ ] Have you search for related issues/PRs?
- [ ] What's the actual output vs expected output?
- [ ] [Optional] Bounty to sponsor the fix ([example](https://www.bountysource.com/issues/66123212-javascript-client-produces-a-wrong-object-for-a-string-enum-type-that-is-used-with-ref))

##### Description

There are quite some errors in the generated Qt5 client code and fail to compile

##### openapi-generator version

4.1.0 and 5.0.x

##### Command line used for generation

java -jar /home/pqy/Downloads/openapi-generator-cli-5.0.0-20190811.165202-11.jar  generate -g c -i https://github.com/jdegre/5GC_APIs/raw/master/TS29502_Nsmf_PDUSession.yaml -o TS29502_Nsmf_PDUSession

### PR DESCRIPTION
**Pull request by etherealjoy**
### PR checklist

- [X] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [X] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh`, `./bin/openapi3/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`. If contributing template-only or documentation-only changes which will change sample output, be sure to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) first.
- [X] Filed the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `4.1.x`, `5.0.x`. Default: `master`.
- [X] Copied the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### Description of the PR
- Add File Upload support for the Qt5 client
- Add File download support for the Qt5 client
- Setup file handling for the Qt5 server only for the common part with client

`ToDo`:
- Setup file handling for the Qt5 server 
- The `HttpFileElement` can be extended to pass `QIODevice` streams instead which includes files, sockets and so on.

`Fixes:`
#3651
#3421
#3457

`Note:`
Not a breaking change for the File handling changes, as it never worked to begin with.

`Sample Usage:`

```C++
#include <QDebug>
#include <QCoreApplication>
#include <QTimer>
#include <OAIDefaultApi.h>

using namespace OpenAPI;

int main(int argc, char *argv[]) {
    QCoreApplication a(argc, argv);
    OAIDefaultApi api;
    api.setHost("http://127.0.0.1:8081");

    OpenAPI::OAIHttpFileElement fileElement;
    fileElement.setMimeType("image/png;charset=UTF-8");
    fileElement.setRequestFileName("FileExt2.png");
    fileElement.setFileName("/home/dev/testapp/ext2.png");
    QString fileId = "fileId";
    qint32 orderId = 555;
    qint32 userId = 777;

    QObject::connect(&api, &OpenAPI::OAIDefaultApi::downLoadFileSignal,
                   [&](OAIHttpFileElement summary)
       {
            qDebug() << "Downloaded the file " << summary.local_filename;
       });

    QObject::connect(&api, &OpenAPI::OAIDefaultApi::upLoadFileSignal,
                   []()
       {
            qDebug() << "Uploaded the file ";
       });

    QObject::connect(&api, &OpenAPI::OAIDefaultApi::upLoadFileWithInfoSignal,
                   []()
       {
            qDebug() << "Uploaded the file with info";
       });
    QTimer::singleShot(0, [&]()
        {
            api.downLoadFile(QString("appsettings.json"));
            api.upLoadFile(fileId, fileElement);
            api.upLoadFileWithInfo(fileId, orderId, userId, fileElement);
        }
    );
    a.exec();
    return 0;
}
```

sample spec for test

```yaml
openapi: 3.0.0
servers:
  - url: 'https://api.testapp.com'
info:
  description: Test API for download and upload features.
  version: 1.0.0
  title: testapp.com
paths:
  /downLoadFile/{fileId}:
    get:
      operationId: downLoadFile
      parameters:
        - name: fileId
          in: path
          schema:
            type: string
          required: true
      responses:
        '200':
          description: A picture file
          content:
            image/png:
              schema:
                type: string
                format: binary

  /upLoadFile/{fileId}:
    post:
      operationId: upLoadFile
      parameters:
        - name: fileId
          in: path
          schema:
            type: string
          required: true
      requestBody:
        content:
          image/png:
            schema:
              type: string
              format: binary          
      responses:
        '200':
          description: success
                
  /upLoadFileWithInfo/{fileId}:
    post:
      operationId: upLoadFileWithInfo
      parameters:
        - name: fileId
          in: path
          schema:
            type: string
          required: true
      requestBody:
        content:
          multipart/form-data:
            schema:
              type: object
              properties:
                orderId:
                  type: integer
                userId:
                  type: integer
                fileName:
                  type: string
                  format: binary          
      responses:
        '200':
          description: success
```

`cc:`
@stkrwork @MartinDelille @muttleyxd @ravinikam

### CODE REVIEW

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/CppQt5AbstractCodegen.java [hunk 1]
```diff
@@ -91,7 +91,7 @@ public CppQt5AbstractCodegen() {
         typeMapping.put("UUID", "QString");
         typeMapping.put("URI", "QString");
         typeMapping.put("file", "QIODevice");
```
Comment (wing328) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/CppQt5AbstractCodegen.java]: @etherealjoy Binary and file are essentially the same so I would suggest mapping `file` to `QByteArray` as well.

Comment (etherealjoy) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/CppQt5AbstractCodegen.java]: Makes, sense.

File: modules/openapi-generator/src/main/resources/cpp-qt5-client/HttpFileElement.cpp.mustache [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    }
+    return QJsonDocument::fromBinaryData(bArray.data()).object();
+}
+
+bool 
+{{prefix}}HttpFileElement::fromStringValue(const QString &instr){
+    QFile file(local_filename);
+    bool result = false;
+    try {
+        if(file.exists()){
```
[the remark this answers is not in the document]
Comment (etherealjoy) [reply modules/openapi-generator/src/main/resources/cpp-qt5-client/HttpFileElement.cpp.mustache]: You are right, none of the `QFile` methods throw an exception

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/CppQt5AbstractCodegen.java (hunk 1 above)
Comment (etherealjoy) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/CppQt5AbstractCodegen.java]: Done

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/CppQt5ClientCodegen.java [hunk 3]
```diff
@@ -74,16 +74,17 @@ public CppQt5ClientCodegen() {
         addSwitch(CodegenConstants.OPTIONAL_PROJECT_FILE, OPTIONAL_PROJECT_FILE_DESC, this.optionalProjectFileFlag);
         supportingFiles.add(new SupportingFile("helpers-header.mustache", sourceFolder, PREFIX + "Helpers.h"));
         supportingFiles.add(new SupportingFile("helpers-body.mustache", sourceFolder, PREFIX + "Helpers.cpp"));
-        supportingFiles.add(new SupportingFile("HttpRequest.h.mustache", sourceFolder, PREFIX + "HttpRequest.h"));
+        supportingFiles.add(new SupportingFile("HttpRequest.h.mustache", sourceFolder, PREFIX + "HttpRequest.h"));
```
[the remark this answers is not in the document]
Comment (etherealjoy) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/CppQt5ClientCodegen.java]: Removed whitespace

File: modules/openapi-generator/src/main/resources/cpp-qt5-client/HttpFileElement.cpp.mustache (hunk 2 above)
[the remark this answers is not in the document]
Comment (etherealjoy) [reply modules/openapi-generator/src/main/resources/cpp-qt5-client/HttpFileElement.cpp.mustache]: Removed exception

**Comment by wing328** [comment]
If no further feedback, I'll merge tomorrow (Tue)

**Pull request merged**